org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/searcher/ArtifactVersionSearcher.java

```diff
@@ -0,0 +1,316 @@
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.index.ArtifactInfo;
+import org.apache.maven.index.ArtifactInfoFilter;
+import org.apache.maven.index.Field;
+import org.apache.maven.index.Indexer;
+import org.apache.maven.index.IteratorSearchRequest;
+import org.apache.maven.index.IteratorSearchResponse;
+import org.apache.maven.index.MAVEN;
+import org.apache.maven.index.context.ExistingLuceneIndexMismatchException;
+import org.apache.maven.index.context.IndexCreator;
+import org.apache.maven.index.context.IndexingContext;
+import org.apache.maven.index.expr.SourcedSearchExpression;
+import org.apache.maven.index.updater.IndexUpdateRequest;
+import org.apache.maven.index.updater.IndexUpdateResult;
+import org.apache.maven.index.updater.IndexUpdater;
+import org.apache.maven.index.updater.ResourceFetcher;
+import org.apache.maven.index.updater.WagonHelper;
+import org.apache.maven.wagon.Wagon;
+import org.apache.maven.wagon.events.TransferEvent;
+import org.apache.maven.wagon.events.TransferListener;
+import org.apache.maven.wagon.observers.AbstractTransferListener;
+import org.codehaus.plexus.PlexusContainer;
+import org.codehaus.plexus.component.repository.exception.ComponentLookupException;
+import org.eclipse.aether.repository.LocalRepositoryManager;
+import org.eclipse.aether.repository.RemoteRepository;
+import org.eclipse.aether.util.version.GenericVersionScheme;
+import org.eclipse.aether.version.InvalidVersionSpecificationException;
+import org.eclipse.aether.version.Version;
+import org.eclipse.lsp4xml.dom.DOMDocument;
+import org.eclipse.lsp4xml.dom.DOMNode;
+import org.eclipse.lsp4xml.extensions.maven.MavenProjectCache;
+import org.eclipse.lsp4xml.extensions.maven.MavenRepositoryCache;
+import org.eclipse.lsp4xml.extensions.maven.VersionValidator;
+
+public class ArtifactVersionSearcher {
+	public static final int COMPARISON_TYPE_GREATER = 1; // 0001
+	public static final int COMPARISON_TYPE_LESS = 2; // 0010
+	public static final int COMPARISON_TYPE_EQUALS = 4; // 0100
+
+	private static final ArtifactVersionSearcher INSTANCE = new ArtifactVersionSearcher();
+
+	org.apache.maven.repository.RepositorySystem repositorySystem;
+
+	LocalRepositoryManager localRepoMan;
+
+	private Indexer indexer;
+
+	private IndexUpdater indexUpdater;
+
+	private Wagon httpWagon;
+
+	private HashMap<String, IndexingContext> indexingContexts = new HashMap<>();
+
+	private List<String> artifactVersions = Collections.synchronizedList(new ArrayList<>());
+
+	private CompletableFuture<Void> syncReq = null;
+
+	private Set<String> brokenContexts = Collections.synchronizedSet(new HashSet<String>());
+
+	private ArtifactVersionSearcher() {
+		this.httpWagon = null;
+		this.indexUpdater = null;
+		this.indexer = null;
+	}
+
+	public static ArtifactVersionSearcher getInstance() {
+		return INSTANCE;
+	}
+
+	public List<String> getArtifactVersions(DOMNode node, MavenProjectCache cache, DOMDocument doc) {
+
```

Comment:
@mickaelistria do you think the way the CompletableFutures are being handled is acceptable for now? Any suggestions on how to improve this function? It has quite a high cognitive complexity according to SonarLint